### PR TITLE
Move reflection based converter and contract resolution logic to DefaultJsonTypeInfoResolver

### PR DESCRIPTION
**Pull request by krwq**
Contributes to: https://github.com/dotnet/runtime/issues/63686

This is preparing code for contract resolvers - moving current reflection resolution logic to a separate class which once APIs are approved will be default contract resolver implementing public interface.

### CODE REVIEW

File: src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/JsonSerializerContextTests.cs [hunk 1]
```diff
@@ -37,18 +37,12 @@ public static void Converters_AndTypeInfoCreator_NotRooted_WhenMetadataNotPresen
...
                     // If any of them changes, this test will need to be kept in sync.
 
-                    // Confirm built-in converters not set.
-                    AssertFieldNull("s_defaultSimpleConverters", optionsInstance: null);
-                    AssertFieldNull("s_defaultFactoryConverters", optionsInstance: null);
-
-                    // Confirm type info dynamic creator not set.
-                    AssertFieldNull("s_typeInfoCreationFunc", optionsInstance: null);
+                    // Confirm reflection resolver not set.
+                    AssertFieldNull("s_reflectionTypeInfoResolver", optionsInstance: null);
```
Comment (krwq) [inline src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/JsonSerializerContextTests.cs:45]: we do not need to check all fields for reflection resolver because it's not possible to create an instance and not root converters

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs [hunk 2]
```diff
@@ -366,17 +229,17 @@ private JsonConverter GetConverterFromAttribute(JsonConverterAttribute converter
...
-                                    // even if the default converters might have been rooted.
-                s_defaultSimpleConverters != null &&
-                s_defaultSimpleConverters.TryGetValue(typeToConvert, out converter))
+            if (_serializerContext != null)
             {
-                return true;
+                // For consistency do not return any default converters for
+                // options instances linked to a JsonSerializerContext,
+                // even if the default converters might have been rooted.
+                converter = null;
```
[the remark this answers is not in the document]
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:237]: In general we don't want to intermix reflection and source-gen metadata during a serialization routine. The most pertinent reason to me personally is so that we don't mask/create bugs in source-gen/AOT scenarios due to reflection providing silent fallback functionality, and make related analysis more difficult (e.g. https://github.com/dotnet/runtime/pull/68464).

> we currently consistently do exactly opposite and use different converter logic

Indeed we fallback to default converters in the `options.GetConverter(Type)` API. This was to prevent breaking changes where we don't return a converter (see https://github.com/dotnet/runtime/pull/51695#discussion_r618841501,  https://github.com/dotnet/runtime/pull/51897), and more importantly because source-gen does not use or need this method. If it did, I'd have advocated for making the breaking change in 6.0.

> I can make sure this code doesn't throw if reflection logic is not initialized

Is this not the eventual behavior when this method returns `null`?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs [hunk 3]
```diff
@@ -601,28 +601,20 @@ internal MemberAccessor MemberAccessorStrategy
 
         /// <summary>
         /// Initializes the converters for the reflection-based serializer.
-        /// <seealso cref="InitializeForReflectionSerializer"/> must be checked before calling.
         /// </summary>
         [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
-        internal void InitializeForReflectionSerializer()
+        [MemberNotNull(nameof(s_reflectionTypeInfoResolver))]
+        internal static void EnsureInitializedForReflectionSerializer()
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:607]: Why the change to `static`?

This was initially changed to a static  in https://github.com/dotnet/runtime/pull/64646 and changed back to an instance method in https://github.com/dotnet/runtime/pull/68496 due to issues that escape me right now. cc @eiriktsarpalis

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Helpers.cs [hunk 4]
```diff
@@ -18,10 +18,7 @@ private static JsonTypeInfo GetTypeInfo(JsonSerializerOptions? options, Type run
             Debug.Assert(runtimeType != null);
 
             options ??= JsonSerializerOptions.Default;
-            if (!options.IsInitializedForReflectionSerializer)
-            {
-                options.InitializeForReflectionSerializer();
-            }
+            JsonSerializerOptions.EnsureInitializedForReflectionSerializer();
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Helpers.cs:21]: I'm guessing the original separation was to avoid making a method call if already initialized -- likely this was overcompensating but wondering if this has any noticeable impact on the benchmark suite.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs [hunk 5]
```diff
@@ -74,10 +74,6 @@ internal void ClearCaches()
         private void InitializeCachingContext()
         {
             _cachingContext = TrackedCachingContexts.GetOrCreate(this);
-            if (IsInitializedForReflectionSerializer)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:77]: Why is this removed? It will break coherency for the shared cache.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs [hunk 6]
```diff
@@ -219,7 +113,7 @@ public JsonConverter GetConverter(Type typeToConvert)
                 ThrowHelper.ThrowArgumentNullException(nameof(typeToConvert));
             }
 
-            RootReflectionSerializerDependencies();
+            EnsureInitializedForReflectionSerializer();
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:116]: Rooting the reflection dependencies is a subset of initializing for the reflection serialization. This is now setting the `IsInitializedForReflectionSerializer` flag when it wasn't doing that in the past, which can result in subtle breaking changes (e.g. when a user calls `GetConverter` in an `Options` instance tied to a `JsonSerializerContext`).

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs (hunk 3 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:607]: My changes unintentionally regressed .NET 6 behavior, namely that `JsonSerializerOptions` instances (including ones emanating from `JsonSerializerContext`) are individually and silently opted into reflection metadata generation when passed into one of the reflection-based serialization methods. Preserving this behaviour necessitates keeping a sentinel value on every options instance.

I'm honestly not a fan of the current behaviour, but if do decide to change it we should be deliberate about it and 1) mark it as a breaking change, 2) expose a knob in `JsonSerializerContext` so that users can easily get back the reflection fallback behavior.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultJsonTypeInfoResolver.cs [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+            }
+        }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "The ctor is marked with RequiresUnreferencedCode.")]
+        internal JsonTypeInfo GetTypeInfo(Type type, JsonSerializerOptions options)
+        {
+            JsonTypeInfo.ValidateType(type, null, null, options);
+
+            MethodInfo methodInfo = typeof(JsonSerializerOptions).GetMethod(nameof(JsonSerializerOptions.CreateReflectionJsonTypeInfo), BindingFlags.NonPublic | BindingFlags.Instance)!;
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultJsonTypeInfoResolver.cs:85]: Why isn't `JsonSerializerOptions.CreateReflectionJsonTypeInfo` moved to this class as well?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultJsonTypeInfoResolver.cs [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+        [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
+        private static void RootReflectionSerializerDependencies()
+        {
+            // s_defaultFactoryConverters is the last field assigned.
+            // Use it as the sentinel to ensure that all dependencies are initialized.
+            if (s_defaultFactoryConverters == null)
+            {
+                s_defaultSimpleConverters = GetDefaultSimpleConverters();
+                s_defaultFactoryConverters = GetDefaultFactoryConverters();
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultJsonTypeInfoResolver.cs:75]: Needs `Volatile.Write` to avoid potential instruction reordering

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/CustomConverterTests/CustomConverterTests.cs [hunk 9]
```diff
@@ -179,40 +179,6 @@ void RunTest<TConverterReturn>()
             }
         }
 
-        [ActiveIssue("https://github.com/dotnet/runtime/issues/66232", TargetFrameworkMonikers.NetFramework)]
-        [ActiveIssue("https://github.com/dotnet/runtime/issues/66371", typeof(PlatformDetection), nameof(PlatformDetection.IsMonoInterpreter))] 
-        [ConditionalFact(typeof(RemoteExecutor), nameof(RemoteExecutor.IsSupported))]
-        public static void GetConverter_Poco_WriteThrowsNotSupportedException()
```
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/CustomConverterTests/CustomConverterTests.cs:185]: This is a regression test validating .NET 6 behaviour. Calling into RUC methods is not equivalent to opting into reflection metadata resolution (though the two concerns often overlap). See also my comments in https://github.com/dotnet/runtime/pull/68496/files#r859817104 and https://github.com/dotnet/runtime/pull/68496/files#r859833346.

As mentioned, it might make sense to change this behavior, but we do need to file a breaking change document and provide knobs for getting the old behavior back in sourcegen. But it's certainly not in scope for a refactoring PR.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultJsonTypeInfoResolver.cs [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Runtime.ExceptionServices;
+using System.Text.Json.Serialization.Converters;
+using System.Threading;
+
+namespace System.Text.Json.Serialization.Metadata
+{
+    internal sealed class DefaultJsonTypeInfoResolver
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultJsonTypeInfoResolver.cs:14]: This is primarily defining a converter resolver and not a `JsonTypeInfo` resolver. Do we need to move all the converter-related code here?

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/CustomConverterTests/CustomConverterTests.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/CustomConverterTests/CustomConverterTests.cs:185]: We don't want a call to `JsonSerializerOptions.GetConverter` to trigger reflection-based metadata resolution for the instance. Most importantly though, it should not cause every other options instance in the process to opt in to reflection-based metadata resolution. That's what the final assertion is meant to check.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs [hunk 11]
```diff
@@ -601,28 +601,20 @@ internal MemberAccessor MemberAccessorStrategy
...
-                _cachingContext.Options.IsInitializedForReflectionSerializer = true;
-            }
+            s_reflectionTypeInfoResolver ??= new DefaultJsonTypeInfoResolver();
         }
 
         private JsonTypeInfo GetJsonTypeInfoFromContextOrCreate(Type type)
         {
             JsonTypeInfo? info = _serializerContext?.GetTypeInfo(type);
-            if (info == null && IsInitializedForReflectionSerializer)
+            if (info == null && s_reflectionTypeInfoResolver != null)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:615]: This will cause every source generating serializer in the process to fall back to reflection just because some unrelated thread ran one reflection-based serialization. Please revert this change.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs (hunk 3 above)
Comment (krwq) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:607]: I understand the current behavior but why do we do that in the first place? Changing from throwing to not throwing is not a breaking change as far as I understand

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (krwq) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs:237]: isn't "intermixing" already contradicting the current logic @eiriktsarpalis describes https://github.com/dotnet/runtime/pull/68496#discussion_r859833346 - that means you can partially use reflection APIs after using one of the unsafe APIs? We should decide one way or another

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs (hunk 3 above)
Comment (krwq) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:607]: let's discuss this during our next feature crew meeting. I'm in favor of making breaking change here (i.e. disallow fallback always)

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:607]: > I understand the current behavior but why do we do that in the first place? Changing from throwing to not throwing is not a breaking change as far as I understand

It doesn't throw because it falls back to the reflection-derived metadata. This might be acceptable for CoreCLR applications, but it invalidates the purpose of source generators.

**Comment by krwq** [comment]
I'll replace this PR shortly for something simpler. I'll not be moving converters to DefaultTypeInfoResolver and will leave that change after we remove fallback logic from SerializerContext (which will require breaking change doc to be created)
